osu.Server.Spectator/Hubs/MultiplayerQueue.cs

```diff
@@ -0,0 +1,315 @@
+            {
+                proposedWereValid = false;
+                foreach (var mod in invalid)
+                    valid.Remove(mod);
+            }
+
+            validMods = valid.Select(m => new APIMod(m));
+
+            return proposedWereValid;
+        }
+
+        /// <summary>
+        /// Ensures that a <see cref="MultiplayerPlaylistItem"/>'s required and allowed mods are compatible with each other and the room's ruleset.
+        /// </summary>
+        /// <param name="item">The <see cref="MultiplayerPlaylistItem"/> to validate.</param>
+        /// <exception cref="InvalidStateException">If the mods are invalid.</exception>
+        private static void ensureModsValid(MultiplayerPlaylistItem item)
+        {
+            // check against ruleset
+            if (!populateValidModsForRuleset(item.RulesetID, item.RequiredMods, out var requiredMods))
+            {
+                var invalidRequiredAcronyms = string.Join(',', item.RequiredMods.Where(m => requiredMods.All(valid => valid.Acronym != m.Acronym)).Select(m => m.Acronym));
+                throw new InvalidStateException($"Invalid mods were selected for specified ruleset: {invalidRequiredAcronyms}");
+            }
+
+            if (!populateValidModsForRuleset(item.RulesetID, item.AllowedMods, out var allowedMods))
+            {
+                var invalidAllowedAcronyms = string.Join(',', item.AllowedMods.Where(m => allowedMods.All(valid => valid.Acronym != m.Acronym)).Select(m => m.Acronym));
+                throw new InvalidStateException($"Invalid mods were selected for specified ruleset: {invalidAllowedAcronyms}");
+            }
+
+            if (!ModUtils.CheckCompatibleSet(requiredMods, out var invalid))
+                throw new InvalidStateException($"Invalid combination of required mods: {string.Join(',', invalid.Select(m => m.Acronym))}");
+
+            // check aggregate combinations with each allowed mod individually.
+            foreach (var allowedMod in allowedMods)
+            {
+                if (!ModUtils.CheckCompatibleSet(requiredMods.Concat(new[] { allowedMod }), out invalid))
+                    throw new InvalidStateException($"Invalid combination of required and allowed mods: {string.Join(',', invalid.Select(m => m.Acronym))}");
+            }
+        }
+
+        /// <summary>
+        /// Verifies all proposed mods are valid for the room's ruleset, returning instantiated <see cref="Mod"/>s for further processing.
+        /// </summary>
+        /// <param name="rulesetID">The legacy ruleset ID to check against.</param>
+        /// <param name="proposedMods">The proposed mods.</param>
+        /// <param name="valid">A list of valid deserialised mods.</param>
+        /// <returns>Whether all <see cref="proposedMods"/> were valid.</returns>
+        private static bool populateValidModsForRuleset(int rulesetID, IEnumerable<APIMod> proposedMods, out List<Mod> valid)
```

Comment:
Ignoring that this `static` method is wedged between non-`static` counterparts, it feels like this shouldn't be here. Did I dream that you were splitting out mod comparison methods like this into a common location? I'd much rather this (and the method above) were in `ModUtils` on the osu! side, or if that doesn't make sense, a counterpart helper class in this project.

Reply:
It's such a specific use-case that I'm not convinced it's worth, but sure.